### ISSUE TITLE
ROCs only applying to planets with the matching template

### ISSUE THREAD

**Issue by Kerbas-ad-astra** [Kopernicus/Kopernicus#397]
I've been starting to mess around with adding ROCs (Breaking Ground planetary features) to the new planets in OPM and RSS, using patches like this:

```
@ROC_DEFINITION:HAS[#Type[LaytheGeyser]]
{
	@displayName = Geyser
	CELESTIALBODY
	{
		Name = Tekto
		Biome = Southern Islands
		Biome = Belial Island
		Biome = Rhat Island
		Biome = Malebolge Peninsula
		Biome = Witvoogd Peninsula
		Biome = Sarbian Islands
		Biome = Tande Region
		Biome = Pozine Region
		Biome = Mammon Region
		Biome = Det Region
		Biome = Adnomimon Region
		Biome = Dis Region
		Biome = Sylandro Region
	}
	CELESTIALBODY
	{
		Name = Eeloo
		Biome = Southern Glaciers
	}
	CELESTIALBODY
	{
		Name = Thatmo
		Biome = Rhat Highlands
		Biome = Felucia Region
		Biome = Ryloth Region
	}
}
```

However, while I could find geysers on Tekto readily enough, they weren't showing up (even with the help of the ROC detector in the cheat menu) on Thatmo, and in fact neither were most of the ROCs I'd configured.  Then I noticed that Tekto was based on the Laythe template, while Eeloo is based on the Eeloo template and Thatmo uses the Moho template, and sure enough, when I added a Thatmo block to the Moho ROC configs, the Moho rocks appeared when the Minmus rocks and Laythe Geysers I had wanted (and configured) did not.  (With and without Kopernicus, I can e.g. add PolYellowStones to Minmus;  it's just the added planets that behave this way.)

I'm doing all of this testing on the 1.8 branch of Kopernicus.

**Comment by Sigma88** [comment]
I don't know if anyone has worked on this, @Kerbas-ad-astra  could you try compiling the 1.8 branch now and check if the issue is still there?

**Comment by Kerbas-ad-astra** [comment]
Yes, it still behaves as before.  I can add any planet's ROCs to any of the other stock planets, but modded planets only get ROCs from their template (e.g. Moho but not Minmus or Laythe for Thatmo).

**Comment by StollD** [comment]
> Yes, it still behaves as before.  I can add any planet's ROCs to any
> of the other stock planets, but modded planets only get ROCs from
> their template (e.g. Moho but not Minmus or Laythe for Thatmo).

It's a design flaw in the ROC system. This fixed it when 1.7 released: 

https://github.com/StollD/MyRocksAreBiggerThanYours

The ROCs dont discover the planet they are attached to themselves, they 
store it as a string internally. When the planet gets cloned that string 
isn't updated, so the ROCs think they are on Laythe instead of Thatmo.

The fix is simple enough that it could be moved into Kopernicus directly. 
I think the only planet mod that ever started using the ROCs was JNSQ so 
it wasn't a priority.

**Comment by Kerbas-ad-astra** [comment]
That did the trick!  I'm feeling pretty good about pushing my configs to OPM, but RSS still has a couple of pinch points that should ideally be resolved before I push it over there:
- Huge planets have huge quads which results in huge performance impacts (~25,000 ROCs being handled at once when landing on Io)...I remember reading a similar issue from rescaled solar systems, though I don't remember if a solution was found.
- I'm not seeing any baobab trees on Earth...I had thought this might be due to some Kerbin/Earth name issue, but even when I cover both bases I'm not seeing any trees after flying halfway across South America:

```
@ROC_DEFINITION:HAS[#Type[KerbinBaobabTree]]
{
	CELESTIALBODY
	{
		Name = Earth
		Biome = Tropics
		Biome = Savanna
	}
	@CELESTIALBODY:HAS[#Name[Kerbin]]
	{
		!Biome,* = delete
		Biome = Tropics
		Biome = Savanna
	}
}
```

**Comment by Sigma88** [comment]
I'll consider adding this feature to Kopernicus if I get some free time, thanks for the help troubleshooting

**Comment by Kerbas-ad-astra** [comment]
An update on putting ROCs on Earth: I've tried increasing the frequency (in case they were being spread thin or something) and commenting out Earth's "cbNameLater" field (it's been known to cause some hiccups in other mods), but neither of those worked.  To be extra-sure I tried to put ROCs on Gael, the homeworld in Galileo's Planet Pack, which is Kerbin-sized and does not use cbNameLater but is a "new" planet that uses Kerbin as its template; I couldn't add ROCs to it, so it's not the size or cbNameLater that are causing trouble but maybe something related to how Kopernicus replaces Kerbin with a new homeworld (I can move Baobab trees to new biomes on Kerbin just fine, even when using the Outer Planets mod).

**Comment by Kerbas-ad-astra** [comment]
It turns out there are two issues.

The first is that the ROC system is extremely brittle; if a ROC_DEFINITION has a CELESTIALBODY block for a planet that doesn't exist, no ROCs show up anywhere.  Since I was starting my RSS config by deleting all CELESTIALBODY nodes and then adding them as I went along, I never had this problem with my RSS install (nor on my OPM install, since it doesn't delete the stock planets), but it was biting me in my testing with GPP because, on that install, I left the stock ROC configs alone; it was only after deleting files and folders bit by bit that I figured out that it was GPP deleting all of the original solar system bodies that was causing trouble.

The second issue is that cbNameLater is bad and kills puppies (ref. KSP-RO/RealSolarSystem#170).  When Earth is configured with cbNameLater = Earth, then Earth exists for the purpose of not breaking all ROCs everywhere in the solar system, but the ROCs don't actually show up on Earth (just all the other planets where I've defined them).  I'm guessing there's some order-of-operations issue at work -- maybe the renaming occurs after ROCs get attached to bodies, so the system tries to put the ROCs on Earth, but fails because there is no Earth because it's still Kerbin, and then by the time Kerbin is renamed, it's too late to get any ROCs.  Likewise, I'm guessing that defining ROCs for Kerbin attaches them to Kerbin in that loading phase, but then Kerbin gets renamed and now the ROCs are gone.

I will compile the 1.9.1 branch and see if any of this has changed, but nothing in the KSP and BG changelogs makes me think it has.

Tl;dr: you can (most likely, pending further testing in KSP 1.9.1) include MyRocksAreBiggerThanYours into mainline Kopernicus, and RSS needs to drop cbNameLater (which will resolve this pinch point and one with WaypointManager, but will require a bunch of configs to be edited to refer internally to Kerbin instead of Earth -- RSSVE for sure, probably RP-0...).

**Comment by Sigma88** [comment]
I'm not sure why MyRocksAreBiggerThanYours was made as an external plugin rather than just added to Kopernicus in the first place, but the fact that it was, makes me think that there was a good reason for it. 

But I am not familiar enough with the ROC system to be of any help here.

I'll need more time to get familiar with it

**Comment by Kerbas-ad-astra** [comment]
Everything works in 1.9.1 as it did in 1.8.1 (MRABTY necessary, ROC_DEFINITION still brittle, cbNameLater still bad for ROCs).

**Comment by R-T-B** [comment]
Closing this old issue as part of a broader cleanup.  If you still have issues reopen a new issue.
